### PR TITLE
add statusBar updates based on cursor

### PR DESCRIPTION
**Pull request by sofroniewn**
# What does this PR do? Why are doing this? References?
This PR adds the x,y mouse coordinates, slicer positions (if any), and the grayscale value of the top visible image layer to the status bar.

Right now the first two axes are labeled x,y and the rest get numbered starting at 2. I can change this default behavior, or we can wait until we revisit axis labels in general.

Right now the first two image indices are assumed to be the ones corresponding to the x,y image, but we will probably need to revisit this once me tackle the more general indexing.

If you move the cursor out of the view of the image it displays the last coordinate positions of the mouse. It could also not display anything, but I prefer this behavior. If you adjust the position of the sliders the values will update too.

Right now it will just ignore any markers layers. I could instead have it show an index for the marker if desired.

Right now it also assumes that there is just one image transform going from the window coordinates to the image coordinates, that of the first layer. If we had images shown next to each other instead of tiling it probably won't work as expected.

Also now it will work best if all image layers have the same dimension. We don't seem to strictly enforce this yet, but I think we're making a number of assumptions that it is the case, or am I wrong?

image

## Type of change
[Please delete options that are not relevant.]
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
[Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.]
- [x] TestCase Run examples/layers.ipynb and interact with the gui

## Final Checklist:
- [x] My PR is the possible minimum work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
Also we don't display the total shape of the images in the viewer anywhere (say the `viewer.max_shape` property). We could put that somewhere too.

Also instead of displaying the update on the statusBar at the bottom of the gui we could add a toolbar along the top of the viewer window instead

**Comment by sofroniewn** [comment]
also right now the grayscale value ignores any opacity settings instead taking the value directly from the top most visible layer. is this behavior undesirable and should it instead reflect the actual pixel value displayed on the screen?

**Comment by sofroniewn** [comment]
I have now added support for rgb images and fixed a bug where I had the x and y coords swapped

**Comment by jni** [comment]
I'm pretty opposed to using x, y — is this for axis 0 and axis 1 or for axis 1 and axis 0? There is a case for either and either is bad for different regions. I agree that axis labeling might need to be solved before we come to a decision here, but in the meantime we can choose something that is at least self-consistent. For example, I think by default the slider axes are the leading axes? In that case, it should be ax0=5, ax1=5, ax3=239, ax4=255, value=.

Other than that, though, this is a wonderful idea and it looks really nice! =) The whole transformations issue requires a more considered response than I have time to write now. =)

**Comment by kevinyamauchi** [comment]
This looks awesome! I'll try to review later today. Thanks, @sofroniewn !

**Comment by royerloic** [comment]
Looks great!
A few comments:
- I would just display a tuple like: (239, 255, 5, 5) 
- The displayed values should correspond to selected layers, if nothing is selected, then top ones are displayed up until a reasonable limit like 5 or so.  The values shoudl really just be coming straight from the layer independently of how it is displayed (transparency, order, etc...)
- We could also display the actual RGB triplet displayed at that pixel, something that could be usefull.
- I like the idea of displaying the point index for point layers.
- Transformations are something that we have to revisit...

File: gui/elements/_viewer.py
Comment (kne42) [inline gui/elements/_viewer.py]:
```python
for i, layer in enumerate(self.layers[::-1]):
    if layer.visible and isinstance(layer, Image):
        top = i
        break
else:
    top = None
```

File: gui/elements/_viewer.py
Comment (kne42) [inline gui/elements/_viewer.py]: This may throw an `IndexError`.

**Review by kne42 — COMMENTED** [review]
Great job Nick! I’ve left a few comments regarding implementation details.

I also agree with Loïc on most of his comments :)

File: gui/elements/_viewer.py
Comment (sofroniewn) [reply gui/elements/_viewer.py]: Looks good, I'll fix this.

File: gui/elements/_viewer.py
Comment (sofroniewn) [reply gui/elements/_viewer.py]: not sure I understand. Is the problem the use of `x`? Let me know any alternatives and I'll change it.

Comment (kne42) [reply gui/elements/_viewer.py]: This will be out of bounds whenever the `viewer.max_shape` is greater than the `layer.shape` or the cursor position is out of bounds of this shape. I would recommend implementing this in a similar fashion to `layer._set_viewer_slice`.

Comment (sofroniewn) [reply gui/elements/_viewer.py]: Ah yes - gets at my question of what are we doing about the cases where not all of the images have the same dimensions - will checkout `layer._set_viewer_slice`

File: gui/elements/_viewer.py
Comment (sofroniewn) [reply gui/elements/_viewer.py]: This didn't quite work as the `i` starts counting from `0` and I want to be indexing into the end of the list, but I made the following change and it works fine

```python
for i, layer in enumerate(self.layers[::-1]):
    if layer.visible and isinstance(layer, Image):
        top = len(self.layers) - 1 - i
        break
    else:
        top = None
```

File: gui/elements/_viewer.py
Comment (sofroniewn) [reply gui/elements/_viewer.py]: It might actually be nice to pull out the bounds checking etc at the beginning of that function into a separate function that I can reuse. I might give that a try.

Comment (sofroniewn) [reply gui/elements/_viewer.py]: I have now done this and pushed the changes, let me know what you think

**Comment by sofroniewn** [comment]
I have switched over to showing a tuple for the coordinates, according to the indexing into the object. That would correspond to (y,x,slider1,slider2) right now. There was another issue #37 about this already, but I’m not sure if we’ve changed anything since then. If we swap things around there I can change things here too.

image

The rgb triplet is now implemented too.

image

I’ll add the index for markers next.

> The displayed values should correspond to selected layers, if nothing is selected, then top ones are displayed up until a reasonable limit like 5 or so. The values shoudl really just be coming straight from the layer independently of how it is displayed (transparency, order, etc...)

I'm not sure about this. I feel like if a layer is selected but not visible, because it is either turned off or hidden underneath other layers it feels strange to me to show its values. I'm thinking analogously to how a color picker tool works in many apps where it just returns the color of what's on the screen irrespective of much else. Similarly if multiple layers are selected (or no layers are selected) it feels strange to now have a list of values there. Open to thoughts from others

**Comment by sofroniewn** [comment]
I have now added functionality to show the index of any given marker that is moused over. In order to be considered the marker needs to belong to a markers layer that is visible and above the top most visible image layer (i.e. it can be seen), otherwise the marker is ignored.

Also right now the indexing into markers goes (x,y, etc.) but into images goes (y,x, etc.). We need to harmonize this somehow.

Finally the marker is deemed under the mouse if the distance from the marker to the mouse is less than twice the marker size in both x,y axes. If the marker is a square this should correspond to exactly the colored part of the marker. If the marker is a circle there will be times when the mouse is outside the circle but inside that bounding box and the index is still shown. Without going down into the vispy layers this is the best I can do. Do people find this behavior objectionable? I looked into render_picking with vispy, and it's certainly possibly but buggy. I imagine we will reuse this marker selection logic when doing image annotations too.

**Comment by sofroniewn** [comment]
image

**Comment by kne42** [comment]
Regarding determining what is on-screen, @royerloic may know some computer graphics techniques that might help us (or are already implemented by `vispy`).

**Comment by sofroniewn** [comment]
Just fixed a transpose error when `sizes` was a list. Also added the layer number in addition to the marker index when hovering over a marker

**Comment by freeman-lab** [comment]
this is really cool! tiny comment about the formatting for values, i find the `r 5, g 4, b 2` hard to read, maybe something like this?

```
(11,61,92) — index 2
(11,61,92) — value 0.5
(11,61,92) — value (5,4,2)
```
or just

```
(11,61,92) — 2
(11,61,92) — 0.5
(11,61,92) — (5,4,2)
```

first is a bit more explicit, probably better

and if possible i'd align them and use fixed decimal places so that the positions of the numbers don't move around while you move -- haven't pulled locally to test, so maybe you're already doing that :)

**Comment by kevinyamauchi** [comment]
@sofroniewn , this is awesome! Super useful. 

> Finally the marker is deemed under the mouse if the distance from the marker to the mouse is less than twice the marker size in both x,y axes. If the marker is a square this should correspond to exactly the colored part of the marker. If the marker is a circle there will be times when the mouse is outside the circle but inside that bounding box and the index is still shown. Without going down into the vispy layers this is the best I can do. Do people find this behavior objectionable? 

I think this is good for now. I think we should use it and see if there are cases where this interaction is ambiguous or confusing.

Also, I agree with @freeman-lab 's comment regarding the display of the value. I like his top suggestion (repeated below for convenience), as I think it is the most clear. Since the value following the coordinate can have multiple meanings, I think including "index" or "value" (or something similar) is useful. 

>(11,61,92) — index 2
(11,61,92) — value 0.5
(11,61,92) — value (5,4,2)

**Comment by sofroniewn** [comment]
ok will get that change to the display implemented using the extra keyword. For markers it will now look like the following as i want to include the layer id. could also do that for the image, but don't yet. Thoughts?

(11,61,92) — index 2, layer 1
(11,61,92) — value 0.5, layer 0
(11,61,92) — value (5,4,2), layer 0

@freeman-lab for the alignment issues, what about for integers - do you want 009 instead of 9?
Or 0009 if there are more than 1000 pixels?
Also for floats do you want 10.43 and 9.441 so fixed number of digits?

**Comment by kevinyamauchi** [comment]
I think that's a great idea to include the layer index. It's so frustrating how it's not always obvious which layer the mouseover values are coming from in ImageJ multiD image stacks in ImageJ!

With regards to padding, do you think we need to add zeros? Can we just have leading spaces? Also, maybe not important for a viewer, but generally, I think it's best to keep the precision fixed. Maybe something like:

```
10.43
 9.44
```

**Comment by sofroniewn** [comment]
I also like fixed precision - leading spaces seem reasonable for the grayscale and rgb values
(65,44,12) - value 123.43
(65,44,12) - value    3.12

I guess you could then do the same for int's too
(65,44,12) - value 123
(65,44,12) - value    3

(though now need to use a font where all characters are same width too ...)

**Comment by sofroniewn** [comment]
hmm github not including a sequence of spaces in front of that 3 ...

**Comment by jni** [comment]
@freeman-lab @sofroniewn is the suggestion to have as many lines in the status bar as there are layers? That seems like it would get real fat real fast?

I'm not at all attached to this idea, but just throwing it out there: maybe the information should be displayed on the layer panel directly?

For numeric values, I'd suggest engineering or engineering-like notation, e.g. "46.7", "3.22", "5.7e6", "87.2e-9" (Alternately, "5.7M", "87.2n".). I'd specifically avoid scientific notation, it drives me up the wall when I have floats in [0, 255] and NumPy helpfully renders it as 5.40000000E+1.

Finally, as to alignment, I don't think we should worry about it in the short term, in the long term we can experiment with various alignment schemes (Qt surely has right-align etc?), and maybe a configuration file for the font, but I wouldn't add complexity to the UI or code before the functionality is in master. I consider hard-setting of the font to be code complexity, and for questionable gain since most monospace fonts are actually less readable anyway.

**Comment by sofroniewn** [comment]
Currently it would only ever show one layer - i think @royerloic may have been suggesting showing more - but I think that would get too complex. Now it just shows info from the top most relevant one.

As to notation with the numbers, I'll def seek to avoid the weird ones. I also don't want to change font at this point either

**Comment by jni** [comment]
Oooh, right, they were different examples stacked on top of one another. Gotcha. =)

**Comment by freeman-lab** [comment]
good call re alignment @jni i wasn't thinking about fonts when i said that :) definitely agree fonts are a whole big bridge we should separately cross as part of other UI / styling work

**Comment by kevinyamauchi** [comment]
+1 for pushing for a useable prototype that communicates the intent of the feature and then performing UX/I testing in the future to optimize!

I vote to merge soon if there aren't outstanding concerns about the implementation.

**Comment by sofroniewn** [comment]
I have now adjusted the text a little to reflect some of the changes and included fixed precision for the floats. I did not do any padding for the integers, in the end I thought it would look stranger than the behavior now where the whole thing just changes length a little when the number of digits changes. Happy to readdress this in the future, but I'd say let's merge now and then we can revisit with the stylings. I just need two approving reviewers!

**Review by kevinyamauchi — APPROVED** [review]
This is awesome! Thanks, Nick! It looks good to me. I would just run it through a linter before merging.

**Review by kne42 — APPROVED** [review]
Will need to address some design choices in the implementation later but it’s good enough to merge for now.

**Comment by sofroniewn** [comment]
Sounds good - some details of the implementation are likely to change in #74 too, so probably best to wait until after that is merged also

**Pull request merged**
